src/metricscollector/collector/apppoller_test.go

```diff
@@ -0,0 +1,191 @@
+							&events.ContainerMetric{
+								ApplicationId: proto.String("test-app-id"),
+								InstanceIndex: proto.Int32(0),
+								MemoryBytes:   proto.Uint64(1234),
+							},
+						}, nil
+					}
+				})
+
+				It("saves all the metrics to database", func() {
+					Expect(database.SaveMetricCallCount()).To(Equal(3))
+				})
+			})
+
+			Context("when all container metrics are empty", func() {
+				BeforeEach(func() {
+					noaa.ContainerMetricsStub = func(appid string, token string) ([]*events.ContainerMetric, error) {
+						Expect(appid).To(Equal("test-app-id"))
+						Expect(token).To(Equal("bearer test-access-token"))
+
+						return []*events.ContainerMetric{}, nil
+					}
+				})
+
+				It("saves nothings to database", func() {
+					Expect(database.SaveMetricCallCount()).To(BeZero())
+				})
+
+			})
+
+			Context("when container metrics are partially empty", func() {
+				BeforeEach(func() {
+					noaa.ContainerMetricsStub = func(appid string, token string) ([]*events.ContainerMetric, error) {
+						Expect(appid).To(Equal("test-app-id"))
+						Expect(token).To(Equal("bearer test-access-token"))
+
+						if noaa.ContainerMetricsCallCount()%2 == 0 {
+							return []*events.ContainerMetric{}, nil
+						} else {
+							return []*events.ContainerMetric{
+								&events.ContainerMetric{
+									ApplicationId: proto.String("test-app-id"),
+									InstanceIndex: proto.Int32(0),
+									MemoryBytes:   proto.Uint64(1234),
+								},
+							}, nil
+						}
+
+					}
+				})
+
+				It("saves non-empty metrics to database", func() {
+					Expect(database.SaveMetricCallCount()).To(Equal(2))
+				})
+			})
+		})
+
+		Context("when retrieving container metrics all fails", func() {
+
+			BeforeEach(func() {
+				noaa.ContainerMetricsReturns(nil, errors.New("an error"))
+			})
+
+			It("saves nothing to database", func() {
+				Expect(database.SaveMetricCallCount()).To(Equal(0))
+			})
+		})
+
+		Context("when retrieving container metrics partially fails", func() {
+			BeforeEach(func() {
+				noaa.ContainerMetricsStub = func(appid string, token string) ([]*events.ContainerMetric, error) {
+					Expect(appid).To(Equal("test-app-id"))
+					Expect(token).To(Equal("bearer test-access-token"))
+
+					if noaa.ContainerMetricsCallCount()%2 == 0 {
+						return nil, errors.New("an error")
+					} else {
+						return []*events.ContainerMetric{
+							&events.ContainerMetric{
+								ApplicationId: proto.String("test-app-id"),
+								InstanceIndex: proto.Int32(0),
+								MemoryBytes:   proto.Uint64(1234),
+							},
+						}, nil
+					}
+				}
+			})
+
+			It("saves successful results to database", func() {
+				Expect(database.SaveMetricCallCount()).To(Equal(2))
+			})
+
+		})
+
+	})
+
+	Describe("Stop", func() {
+		JustBeforeEach(func() {
+			poller.Stop()
+		})
+
+		Context("when apppoller is started", func() {
+			BeforeEach(func() {
+				poller.Start()
+				time.Sleep(2500 * time.Millisecond)
```

Comment:
Still wrong. It makes little sense. You have no guarantee the loop ever ran or will run.
Normally, you would start, wait for a metricscallcount == 1 to verify the poller is running and then stop it and make sure it never happens. Again it is a weak test since it is time dependent which means you should also adjust the poll interval to make it happen very often like 5ms or 10ms.